File: backend/app/schemas/virtualization.py
"""Virtualization (VMware) inventory API contracts (W1-T3): read-only, no write path.

Mirrors :mod:`app.schemas.devices`: read models map straight off the
``Normalized*Row`` ORM rows in :mod:`app.models.virtualization` via
``from_attributes``. Nested ``nics``/``pnics`` are JSON lists of plain dicts;
Pydantic converts each dict into :class:`VirtualNicRead`/:class:`PhysicalNicRead`
the same way ``from_attributes`` converts scalar columns.

Port groups store the optional natural-key columns (``datacenter``,
``host_name``, ``moref``) as the ``''`` sentinel at the DB layer (NULL would
silently disable the unique constraint — see ``app.models.inventory`` module
docstring); :func:`_empty_to_none` maps ``''`` back to ``None`` on the read
side, the same pattern as ``DeviceNeighborRead.neighbor_interface``.
"""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Annotated

from pydantic import BaseModel, BeforeValidator, ConfigDict

from app.schemas.normalized import HostConnectionState, VirtualSwitchType, VmPowerState

__all__ = [
    "ComputeClusterListResponse",
    "ComputeClusterRead",
    "HypervisorHostListResponse",
    "HypervisorHostRead",
    "PhysicalNicRead",
    "PortGroupListResponse",
    "PortGroupRead",
    "VirtualMachineListResponse",
    "VirtualMachineRead",
    "VirtualNicRead",
]


def _empty_to_none(value: object) -> object:
    """Map the ``''`` natural-key sentinel back to ``None`` (read direction)."""
    return None if value == "" else value


EmptyToNone = Annotated[str | None, BeforeValidator(_empty_to_none)]


class VirtualNicRead(BaseModel):
    """One VM vNIC, nested inside :class:`VirtualMachineRead` (ADR-0051 §5.3)."""

    model_config = ConfigDict(from_attributes=True)

    label: str
    mac_address: str
    port_group_name: str | None
    switch_type: VirtualSwitchType | None
    connected: bool
    ip_addresses: list[str]


class VirtualMachineRead(BaseModel):
    """One virtual machine (``GET /virtualization/vms[/{id}]``)."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    device_id: uuid.UUID
    name: str
    moref: str
    instance_uuid: str | None
    is_template: bool
    power_state: VmPowerState
    guest_hostname: str | None
    guest_ip_addresses: list[str]
    host_name: str | None
    cluster_name: str | None
    datacenter: str | None
    nics: list[VirtualNicRead]
    description: str | None
    collected_at: datetime
    source_vendor: str


class VirtualMachineListResponse(BaseModel):
    """Paginated VM collection (``GET /virtualization/vms``)."""

    items: list[VirtualMachineRead]
    total: int
    limit: int
    offset: int


class PhysicalNicRead(BaseModel):
    """One host pNIC, nested inside :class:`HypervisorHostRead` (ADR-0051 §5.3)."""

    model_config = ConfigDict(from_attributes=True)

    name: str
    mac_address: str
    link_speed_mbps: int | None


class HypervisorHostRead(BaseModel):
    """One hypervisor host (``GET /virtualization/hosts[/{id}]``)."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    device_id: uuid.UUID
    name: str
    moref: str
    cluster_name: str | None
    datacenter: str | None
    vendor: str | None
    model: str | None
    hypervisor_version: str | None
    connection_state: HostConnectionState
    in_maintenance_mode: bool
    management_ip: str | None
    pnics: list[PhysicalNicRead]
    collected_at: datetime
    source_vendor: str


class HypervisorHostListResponse(BaseModel):
    """Paginated host collection (``GET /virtualization/hosts``)."""

    items: list[HypervisorHostRead]
    total: int
    limit: int
    offset: int


class ComputeClusterRead(BaseModel):
    """One compute cluster (``GET /virtualization/clusters[/{id}]``)."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    device_id: uuid.UUID
    name: str
    moref: str
    datacenter: str | None
    drs_enabled: bool | None
    ha_enabled: bool | None
    collected_at: datetime
    source_vendor: str


class ComputeClusterListResponse(BaseModel):
    """Paginated cluster collection (``GET /virtualization/clusters``)."""

    items: list[ComputeClusterRead]
    total: int
    limit: int
    offset: int


class PortGroupRead(BaseModel):
    """One port group (``GET /virtualization/port-groups[/{id}]``).

    ``datacenter``/``host_name``/``moref`` map the ``''`` sentinel back to
    ``None`` (standard port groups have no ``moref``; distributed port groups
    have no ``host_name``).
    """

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    device_id: uuid.UUID
    name: str
    switch_name: str
    switch_type: VirtualSwitchType
    datacenter: EmptyToNone
    host_name: EmptyToNone
    vlan_id: int | None
    moref: EmptyToNone
    uplink_pnic_names: list[str]
    collected_at: datetime
    source_vendor: str


class PortGroupListResponse(BaseModel):
    """Paginated port-group collection (``GET /virtualization/port-groups``)."""

    items: list[PortGroupRead]
    total: int
    limit: int
    offset: int

File: frontend/src/api/virtualization.ts
/**
 * Typed client functions for the virtualization (VMware) inventory endpoints.
 *
 * Mirrors ``src/api/devices.ts``: the backend schemas in
 * ``app/schemas/virtualization.py`` and the routes in
 * ``app/api/v1/virtualization.py`` (W1-T3). Read-only — there is no write path.
 */

import { apiFetch } from "./client";

// ── Enums (match backend Vm*/Host*/VirtualSwitchType StrEnum values) ─────────

export type VmPowerState = "powered_on" | "powered_off" | "suspended" | "unknown";
export type HostConnectionState = "connected" | "disconnected" | "not_responding" | "unknown";
export type VirtualSwitchType = "standard" | "distributed";

// ── Response shapes ───────────────────────────────────────────────────────────

export interface VirtualNicRead {
  label: string;
  mac_address: string;
  port_group_name: string | null;
  switch_type: VirtualSwitchType | null;
  connected: boolean;
  ip_addresses: string[];
}

export interface VirtualMachineRead {
  id: string;
  device_id: string;
  name: string;
  moref: string;
  instance_uuid: string | null;
  is_template: boolean;
  power_state: VmPowerState;
  guest_hostname: string | null;
  guest_ip_addresses: string[];
  host_name: string | null;
  cluster_name: string | null;
  datacenter: string | null;
  nics: VirtualNicRead[];
  description: string | null;
  collected_at: string;
  source_vendor: string;
}

export interface VirtualMachineListResponse {
  items: VirtualMachineRead[];
  total: number;
  limit: number;
  offset: number;
}

export interface PhysicalNicRead {
  name: string;
  mac_address: string;
  link_speed_mbps: number | null;
}

export interface HypervisorHostRead {
  id: string;
  device_id: string;
  name: string;
  moref: string;
  cluster_name: string | null;
  datacenter: string | null;
  vendor: string | null;
  model: string | null;
  hypervisor_version: string | null;
  connection_state: HostConnectionState;
  in_maintenance_mode: boolean;
  management_ip: string | null;
  pnics: PhysicalNicRead[];
  collected_at: string;
  source_vendor: string;
}

export interface HypervisorHostListResponse {
  items: HypervisorHostRead[];
  total: number;
  limit: number;
  offset: number;
}

export interface ComputeClusterRead {
  id: string;
  device_id: string;
  name: string;
  moref: string;
  datacenter: string | null;
  drs_enabled: boolean | null;
  ha_enabled: boolean | null;
  collected_at: string;
  source_vendor: string;
}

export interface ComputeClusterListResponse {
  items: ComputeClusterRead[];
  total: number;
  limit: number;
  offset: number;
}

export interface PortGroupRead {
  id: string;
  device_id: string;
  name: string;
  switch_name: string;
  switch_type: VirtualSwitchType;
  datacenter: string | null;
  host_name: string | null;
  vlan_id: number | null;
  moref: string | null;
  uplink_pnic_names: string[];
  collected_at: string;
  source_vendor: string;
}

export interface PortGroupListResponse {
  items: PortGroupRead[];
  total: number;
  limit: number;
  offset: number;
}

// ── Query-string params ───────────────────────────────────────────────────────

export interface ListVirtualMachinesParams {
  device_id?: string;
  power_state?: VmPowerState;
  cluster_name?: string;
  limit?: number;
  offset?: number;
}

export interface ListHypervisorHostsParams {
  device_id?: string;
  connection_state?: HostConnectionState;
  cluster_name?: string;
  limit?: number;
  offset?: number;
}

export interface ListComputeClustersParams {
  device_id?: string;
  datacenter?: string;
  limit?: number;
  offset?: number;
}

export interface ListPortGroupsParams {
  device_id?: string;
  switch_type?: VirtualSwitchType;
  limit?: number;
  offset?: number;
}

// ── API functions ─────────────────────────────────────────────────────────────

function buildQuery<T extends object>(params: T): string {
  const qs = new URLSearchParams();
  for (const [key, value] of Object.entries(params as Record<string, string | number | undefined>)) {
    if (value !== undefined) qs.set(key, String(value));
  }
  const query = qs.toString();
  return query ? `?${query}` : "";
}

/** ``GET /api/v1/virtualization/vms`` — paginated, filterable VM list. */
export function listVirtualMachines(
  params: ListVirtualMachinesParams = {},
): Promise<VirtualMachineListResponse> {
  return apiFetch<VirtualMachineListResponse>(`/virtualization/vms${buildQuery(params)}`);
}

/** ``GET /api/v1/virtualization/vms/{id}`` — one VM by id. */
export function getVirtualMachine(id: string): Promise<VirtualMachineRead> {
  return apiFetch<VirtualMachineRead>(`/virtualization/vms/${id}`);
}

/** ``GET /api/v1/virtualization/hosts`` — paginated, filterable host list. */
export function listHypervisorHosts(
  params: ListHypervisorHostsParams = {},
): Promise<HypervisorHostListResponse> {
  return apiFetch<HypervisorHostListResponse>(`/virtualization/hosts${buildQuery(params)}`);
}

/** ``GET /api/v1/virtualization/hosts/{id}`` — one host by id. */
export function getHypervisorHost(id: string): Promise<HypervisorHostRead> {
  return apiFetch<HypervisorHostRead>(`/virtualization/hosts/${id}`);
}

/** ``GET /api/v1/virtualization/clusters`` — paginated, filterable cluster list. */
export function listComputeClusters(
  params: ListComputeClustersParams = {},
): Promise<ComputeClusterListResponse> {
  return apiFetch<ComputeClusterListResponse>(`/virtualization/clusters${buildQuery(params)}`);
}

/** ``GET /api/v1/virtualization/clusters/{id}`` — one cluster by id. */
export function getComputeCluster(id: string): Promise<ComputeClusterRead> {
  return apiFetch<ComputeClusterRead>(`/virtualization/clusters/${id}`);
}

/** ``GET /api/v1/virtualization/port-groups`` — paginated, filterable port-group list. */
export function listPortGroups(
  params: ListPortGroupsParams = {},
): Promise<PortGroupListResponse> {
  return apiFetch<PortGroupListResponse>(`/virtualization/port-groups${buildQuery(params)}`);
}

/** ``GET /api/v1/virtualization/port-groups/{id}`` — one port group by id. */
export function getPortGroup(id: string): Promise<PortGroupRead> {
  return apiFetch<PortGroupRead>(`/virtualization/port-groups/${id}`);
}
